### PR TITLE
Add a new visitor: `SyntaxTransformVisitor`.

### PR DESCRIPTION
**Pull request by zoecarver**
This visitor returns a transformed Syntax node rather than exposing a pre-post visitor. This will be helpful for ASTGen.

### CODE REVIEW

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb [hunk 1]
```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
     }
   }
 }
+
+public protocol SyntaxTransformVisitor {
+  associatedtype ResultType = Void
```
Comment (zoecarver) [inline Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:146]: @CodaFi we had discussed having one of these for each kind of thing we might want to transform (decls, exprs, ec.), but I think one is OK for now. If the need arises to have different ones, we can update the visitor.

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb [hunk 2]
```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
...
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: nil by default.
+  public func visit(_ node: ${node.name}) -> ResultType? {
+%   if node.is_base():
+      // Avoid calling into visitChildren if possible.
+      if !node.raw.layoutView!.children.isEmpty {
+        return visitChildren(node).first
```
Comment (zoecarver) [inline Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:174]: Oops, fixing this.

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb [hunk 3]
```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
...
+    case .${node.swift_syntax_kind}:
+      let node = ${node.name}(data)
+      return visit(node)
+  % end
+    }
+  }
+
+  public func visitChildren<SyntaxType: SyntaxProtocol>(_ node: SyntaxType) -> [ResultType] {
+    let syntaxNode = Syntax(node)
+        return NonNilRawSyntaxChildren(syntaxNode, viewMode: .sourceAccurate).compactMap { childRaw in
```
Comment (zoecarver) [inline Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:210]: Should `viewMode` be an input?

File: Sources/SwiftSyntax/SyntaxData.swift [hunk 4]
```diff
@@ -213,7 +213,7 @@ class SyntaxBox: CustomStringConvertible,
 ///
 /// SyntaxData is an implementation detail, and should not be exposed to clients
 /// of SwiftSyntax.
-struct SyntaxData {
+public struct SyntaxData {
```
Comment (CodaFi) [inline Sources/SwiftSyntax/SyntaxData.swift:216]: Let's not expose this. Take is as `@_spi(RawSyntax)` if you need this.

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb [hunk 5]
```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
...
+      if !node.raw.layoutView!.children.isEmpty {
+        return visitChildren(node)
+      }
+      return []
+%   end
+  }
+%   end
+% end
+
+  public func visit(_ data: SyntaxData) -> [ResultType] {
```
Comment (CodaFi) [inline Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:188]: We don't want clients visiting SyntaxData directly.

Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:188]: I agree, `SyntaxTransformVisitor` should be a `class` and this should be `private` like we do in all the other initializers.

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb [hunk 6]
```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
...
+    // ~50KB per call to this function. rdar://55929175
+  % for node in NON_BASE_SYNTAX_NODES:
+    case .${node.swift_syntax_kind}:
+      let node = ${node.name}(data)
+      return visit(node)
+  % end
+    }
+  }
+
+  public func visitChildren<SyntaxType: SyntaxProtocol>(_ node: SyntaxType) -> [ResultType] {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:208]: Similarly to `visit(_ data: SyntaxData)`, this should be `private`.

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb [hunk 7]
```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
     }
   }
 }
+
+public protocol SyntaxTransformVisitor {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:145]: Let’s put this in a different file. `SyntaxRewriter` also has its dedicated file.

**Review by ahoppen — COMMENTED** [review]
I’m a little worried about this implementation because IIUC its default implementation discards any recursive transformations by default. My suggestion would be to have a `visitAny` method (similar to `SyntaxAnyVisitor`) that all the `visit` methods default to and have that `visitAny` method `fatalError` by default. This way you’ll trap at runtime rather than just dropping results.

Also, I’m not a huge fan of tying this to `Array` as the transformation result. What do you think about having two generic types: `NodeTransformResult` and `TransformResult` (feel free to come up with a better name) and then combining them using a `func reduce(initialResult: TransformResult?, nextPartialResult: NodeTransformResult)`. That way you could also return strings `testStringify` instead of arrays of strings.

**Comment by ahoppen** [comment]
rdar://99492982

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb (hunk 5 above)
Comment (zoecarver) [reply Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:188]: Why do we want this to be a class? I think this should just take a `Syntax`.

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb (hunk 6 above)
Comment (zoecarver) [reply Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:208]: Why? It is useful to API consumers (see the FuncCounter test).

**Comment by zoecarver** [comment]
> I’m a little worried about this implementation because IIUC its default implementation discards any recursive transformations by default. My suggestion would be to have a visitAny method (similar to SyntaxAnyVisitor) that all the visit methods default to and have that visitAny method fatalError by default. This way you’ll trap at runtime rather than just dropping results.

I'm not sure what you mean by this. The default implementation will visit all the children and flatMap them together. Nothing should be lost...

> Also, I’m not a huge fan of tying this to Array as the transformation result. What do you think about having two generic types: NodeTransformResult and TransformResult (feel free to come up with a better name) and then combining them using a func reduce(initialResult: TransformResult?, nextPartialResult: NodeTransformResult). That way you could also return strings testStringify instead of arrays of strings.

As long as there can be N children, I think this implementation will be hard. For a generic implementation, the reducer will either have to drop something on the floor or be non-trivial. I like that with the array approach we can handle any input and pass it up to the client for processing/validation. 

If performance is a concern, maybe we can introduce a SingleTransformVisitor that returns a `T?` instead of `[T]` and asserts that all nodes have one or no children. Otherwise, maybe we can work on your suggested implementation offline or commit it as a follow-up change?

**Comment by zoecarver** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
@swift-ci test

**Comment by zoecarver** [comment]
@swift-ci test

**Comment by zoecarver** [comment]
The `visitAny` method was a great idea. The stringify test did not change, but the FuncCounter got a lot simpler :) Thanks, @ahoppen!

File: Sources/SwiftSyntax/SyntaxTransform.swift.gyb [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  public func visit(_ token: TokenSyntax) -> ResultType { visitAny(Syntax(token)) }
+  public func visit(_ node: UnknownSyntax) -> ResultType { visitAny(Syntax(node)) }
+
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: nil by default.
+  public func visit(_ node: ${node.name}) -> ResultType {
+      visitAny(Syntax(node))
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxTransform.swift.gyb:48]: Nitpick
```suggestion
    visitAny(Syntax(node))
```

**Comment by zoecarver** [comment]
@swift-ci please test

File: Sources/SwiftSyntax/SyntaxTransform.swift.gyb [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  }
+%   end
+% end
+
+  public func visit(_ data: Syntax) -> ResultType {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxTransform.swift.gyb:53]: Nit: Now that this is no longer `SyntaxData`, the parameter should be named `node`.

File: Sources/SwiftSyntax/SyntaxTransform.swift.gyb [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    switch data.raw.kind {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxTransform.swift.gyb:54]: You should be able to use `as(SyntaxEnum.self)` here. That would eliminate the need for force unwraps.

Something like
```swift
    switch node.as(SyntaxEnum.self) {
    case .token(let node):
      return visit(node)
    case .unknown(let node):
      return visit(node)
% for node in NON_BASE_SYNTAX_NODES:
    case .${node.swift_syntax_kind}(let node):
      return visit(node)
% end
```

File: Sources/SwiftSyntax/SyntaxTransform.swift.gyb [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  %   if node.base_kind == "Expr":
+    case .${node.swift_syntax_kind}:
+      let node = data.as(${node.name}.self)!
+      return visit(node)
+  %   end
+  % end
+    default:
+      fatalError("Not expression?")
+    }
+  }
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxTransform.swift.gyb:85]: Why do we need this method? I don’t think we should ever end up calling `visit` with a base node type. Same for the methods below.

File: Tests/SwiftParserTest/SyntaxTransformVisitor.swift [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+      
+      public func visitType(_ paramType: TypeSyntaxProtocol) -> String {
+        if let id = SimpleTypeIdentifierSyntax(Syntax(fromProtocol: paramType)) {
+          return visit(id)
+        }
+        if let arr = ArrayTypeSyntax(Syntax(fromProtocol: paramType)) {
+          return visit(arr)
+        }
+        return "<Unknown>"
+      }
```
Comment (ahoppen) [inline Tests/SwiftParserTest/SyntaxTransformVisitor.swift:55]: Do we even need this. I think `visit(Syntax)` should directly dispatch to `visit(SimpleTypeIdentifierSyntax)` or `visit(ArrayTypeSyntax)`.

File: Tests/SwiftParserTest/SyntaxTransformVisitor.swift [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+      
+      public func visit(_ node: SimpleTypeIdentifierSyntax) -> String {
+        node.name.text
+      }
+
+      public func visit(_ node: ArrayTypeSyntax) -> String {
+          "[" + visitType(node.elementType) + "]"
+      }
+    }
+    XCTAssertNoThrow(try {
```
Comment (ahoppen) [inline Tests/SwiftParserTest/SyntaxTransformVisitor.swift:87]: You don’t need `XCTAssertNoThrow` if you make the test case throwing. IMHO it just reads better without XCTAssertNoThrow.

File: Tests/SwiftParserTest/SyntaxTransformVisitor.swift [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+              }
+            }
+          }
+          """)
+      let count = FuncCounter().visit(parsed)
+      XCTAssertEqual(count, 4)
+    }())
+  }
+  
+  public func testStringify() {
```
Comment (ahoppen) [inline Tests/SwiftParserTest/SyntaxTransformVisitor.swift:41]: `testStringify` is a somewhat misleading name for this test IMO. What do you think about `testPrintFunctionType`

File: Tests/SwiftParserTest/SyntaxTransformVisitor.swift [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    struct Stringify: SyntaxTransformVisitor {
```
Comment (ahoppen) [inline Tests/SwiftParserTest/SyntaxTransformVisitor.swift:42]: Similarly to the test name, what do you think about renaming this:

```suggestion
    /// Assuming there is a single functioned declaration in the source file, print the type of that function.
    struct PrintFunctionType: SyntaxTransformVisitor {
```

**Comment by zoecarver** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
@swift-ci please test

**Pull request merged**